wezterm-gui/src/overlay/copy.rs

```diff
@@ -1141,13 +1155,37 @@ impl Pane for CopyOverlay {
                 (KeyCode::Char(c), KeyModifiers::NONE)
                 | (KeyCode::Char(c), KeyModifiers::SHIFT) => {
                     // Type to add to the pattern
-                    render.pattern.push(c);
+                    if render.pattern.capacity() - render.pattern.len() == 0 {
+                        render.pattern.reserve(10);
+                    }
+                    let position = render.search_cursor.x - SEARCH_CURSOR_PADDING;
+                    render.pattern.insert(position, c);
+                    render.search_cursor.x += 1;
+
                     render.schedule_update_search();
                 }
                 (KeyCode::Backspace, KeyModifiers::NONE) => {
                     // Backspace to edit the pattern
-                    render.pattern.pop();
-                    render.schedule_update_search();
+                    if render.pattern.len() > 0
+                        && render.search_cursor.x - SEARCH_CURSOR_PADDING > 0
+                    {
+                        let position = render.search_cursor.x - SEARCH_CURSOR_PADDING;
+                        render.pattern.remove(position - 1);
+                        if render.search_cursor.x - SEARCH_CURSOR_PADDING > 0 {
+                            render.search_cursor.x -= 1;
+                        }
+                        render.schedule_update_search();
+                    }
+                }
+                (KeyCode::LeftArrow, KeyModifiers::NONE) => {
+                    if render.search_cursor.x - SEARCH_CURSOR_PADDING > 0 {
+                        render.search_cursor.x -= 1;
```

Comment:
This is the hard part: logically, we want to move the cursor one column position but since the `pattern` field is a UTF-8 string, it is a multi-byte unicode sequence, and the required logic is much more complex than this simple subtraction.

Take a look at https://github.com/wez/wezterm/blob/f5e496ebc233187c076cbb34b3cd0f0c50290f51/termwiz/src/lineedit/mod.rs#L518-L529 for how line editing is handled in the termwiz crate.

What I'd prefer to happen here is to refactor that termwiz line editor core state so that the `line` and `cursor` fields are moved into a separate struct, maybe named something like `LineEditBuffer` which has the methods for editing the buffer and moving the cursor.

Then both the line editor in termwiz and overlays like this in wezterm can re-use the same underlying editor logic and it becomes pretty easy to expose more powerful editing options here.

Reply:
Yes, I see. The idea with `LineEditBuffer` sounds good. Then I'll add this to the pr.